DolphinQt: Avoid leaking the GameListModel instance to gracefully shutdown the GameTracker and prevent a crash on exit

cristian64 opened the pull request:
Because the **GameListModel** was never destroyed, its **GameTracker** instance outlived the **Config** module, which led to some crashes on shutdown.

The fix consists of binding the lifespan of **GameListModel** to the **GameList** instance, ensuring that the **GameTracker** is indeed destructed.

In order to be able to destroy the **GameTracker** instance, some _race conditions_, _deadlocks_ and _other crashes_ had to be addressed. These issues were all caused by a bad combination of the **GameTracker**'s worker thread and the `RunOnObject()` function.

To summarize the list of changes:
- All references to the **GameListModel** have been removed from **Settings**.
- The **GameListModel** instance is now owned by the **GameList** instance.
- Components that need to access the **GameListModel** instance are provided with a const reference in their constructors.
- The **GameTracker**'s internal worker thread is now flushed on shutdown, to avoid it accessing stale data, and to favor responsiveness.
- `RunOnObject()` has been replaced with `QueueOnObject()` in **GameTracker**, to prevent a deadlock on shutdown.
- (Extra) **Refresh** button and **Purge Game List Cache** menu entry have now their enabled state synced correctly, and they are not much more responsive.

---

The crash could be reproduced with the following steps:
- Open **Dolphin**.
- Start copying (or downloading) a game into any of the directories monitored by **Dolphin**. This will make the game list to be refreshed every now and then, until the game is fully copied (or downloaded).
- Exit **Dolphin** by clicking on the close button, or selecting **File > Exit** in the top bar menu.
- The application will potentially crash on shutdown:

```
Thread 10 "dolphin-emu" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffc59f6700 (LWP 6006)]
0x000055555566b41b in bool Config::Get<bool>(Config::ConfigInfo<bool> const&) ()
(gdb) bt
#0  0x000055555566b41b in bool Config::Get<bool>(Config::ConfigInfo<bool> const&) ()
#1  0x00005555559a1c3f in UICommon::GameFile::CustomCoverChanged() ()
#2  0x00005555559a8442 in UICommon::GameFileCache::UpdateAdditionalMetadata(std::shared_ptr<UICommon::GameFile>*) [clone .constprop.227] ()
#3  0x00005555559a9363 in UICommon::GameFileCache::AddOrGet(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool*) ()
#4  0x000055555571bd1b in GameTracker::LoadGame(QString const&) [clone .part.191] ()
#5  0x000055555571e08b in GameTracker::UpdateFileInternal(QString const&) ()
#6  0x000055555571fb8d in std::_Function_handler<void (GameTracker::Command), GameTracker::Command(QObject*)::{lambda(GameTracker::Command)#3}>::_M_invoke(std::_Any_data const&, GameTracker::Command&&) ()
#7  0x00005555557200a8 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<Common::WorkQueueThread<GameTracker::Command>::Reset(std::function<void (GameTracker::Command)>)::{lambda()#1}> > >::_M_run() ()
#8  0x00007ffff09eb57f in  () at /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#9  0x00007ffff0cbe6db in start_thread (arg=0x7fffc59f6700) at pthread_create.c:463
#10 0x00007ffff02c088f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

Techjar commented on Source/Core/DolphinQt/GameList/GameList.cpp:
```diff
@@ -53,7 +53,9 @@ static bool CompressCB(const std::string&, float, void*);
 GameList::GameList(QWidget* parent) : QStackedWidget(parent)
 {
-  m_model = Settings::Instance().GetGameListModel();
+  m_model = new GameListModel(this);
```
`m_model` should also use `shared_ptr` to ensure GameListModel is not deleted at least until GameList itself is deleted.

cristian64 replied:
As mentioned above, QPointer has been used instead, so that the GameListModel ownership is then well defined by the GameList.

cristian64 commented on Source/Core/DolphinQt/NetPlay/NetPlaySetupDialog.cpp:
```diff
@@ -280,17 +279,22 @@ void NetPlaySetupDialog::PopulateGameList()
-  for (int i = 0; i < m_game_list_model->rowCount(QModelIndex()); i++)
+
+  auto* const game_list_model = Settings::Instance().GetGameListModel();
+  if (game_list_model)
```
Well spotted.

cristian64 commented:
@Techjar Thanks for the feedback! Much appreciated.

I wasn't too sure about whether using shared pointers was a good fit for this case. Since these are all UI components, I assumed (maybe incorrectly) that all this would be running in the UI thread, therefore using shared pointers to prevent race conditions would be an overkill.

As of this pull request, the **GameListModel** is constructed with **GameList** as a parent, so it lives just about as long as the **GameList** instance. However, it relies on `GameList::~GameList()` to reset the pointer via `Settings::SetGameListModel(nullptr)`, which feels pretty weak.

By using a combination of *std::shared_ptr* and *std::weak_ptr*, we can still make the **GameList** own the **GameListModel** instance, and whilst the pointer in **Settings** would be auto reset when the **GameList** instance is destroyed. I will amend the pull request later today.

Another good question... Why is **Settings** in charged of storing the **GameListModel** single instance pointer (or, after this pull request, a pointer to the single instance)? Thoughts?

BhaaLseN commented:
I don't see any reason why it is actually stored in `Settings`. It feels like they should be passed around from something at the top (preferably using constructors) instead of having every component grab it from somewhere else.

cristian64 commented:
Following @BhaaLseN 's idea, I ended up removing the cheeky `Settings::GetGameListModel()` function. Now the **GameListModel** reference is passed in in the constructors of those components that need it.

However (1), after doing so, I've then discovered that there was probably a reason why it was _decided_ to leak the **GameListModel/GameTracker** instances... Basically, if the reproduction steps are followed (as described in the pull request's description), there will be a big chance of deadlock when the **GameTracker**'s destructor is executed. I've managed to get around the deadlock by calling `QCoreApplication::processEvents()`. There is a commit for it, where I attempted to describe the issue in depth (if you have not looked at that commit, yet, do it now, please, so that the following makes sense).

However (2), this work-around has only surfaced more issues... The code running inside the `RunOnObject()` may try to access some of the **GameTracker**'s ancestors (such as the **GameListModel**), which have already been destroyed, and will cause another crash.

More work is required... My feeling is that both `RunOnObject()` and the **GameTracker**'s worker thread need to be reconsidered. I don't think `RunOnObject()` is required; a combination of signal/slots should be used instead. When it comes to the worker thread, we'll see.

cristian64 commented on Source/Core/DolphinQt/GameList/GameList.cpp:
```diff
@@ -347,26 +347,25 @@ void GameList::ShowContextMenu(const QPoint&)
-          model->AddGameTag(path, tag);
-      });
+      connect(tag_action, &QAction::toggled, this,
+              [ this, path, tag, model = m_model ](bool checked) {
```
No need to capture `m_model` in a separate parameter. `this` is already available, `m_model` can be used instead.

cristian64 commented:
> I also think it would be better to use callbacks (or Qt signals/slots) in the classes that need access to the model. As far as I could tell, they only need the model to get a list of all games; and for that they could simply raise a signal to have someone else fill in the blanks. Could look like this (somewhat pseudo, assuming it works similarly to events in C#; at least docs say that code after `emit` runs once all the slots have finished):
> 
> ```c++
> // in NetPlaySetupDialog (and the other places):
> signals:
> void RequestGameList(std::vector<GameInfo>& list_of_games_to_be_shown);
> 
> public:
> void PopulateGameList()
> {
>   std::vector<GameInfo> list_of_games_to_show_here;
>   emit RequestGameList(list_of_games_to_show_here);
>   for (auto game : list_of_games_to_show_here) { /* ... */ }
> }
> 
> // in GameList (or any other place that might be appropriate):
> slots:
> void FillRequestedGameList(std::vector<GameInfo>& list_of_games_to_be_shown)
> {
>   // add all the games that should be shown, potentially filtering them first
>   list_of_games_to_be_shown.push_back(game1);
>   list_of_games_to_be_shown.push_back(game2);
>   list_of_games_to_be_shown.push_back(game7);
> }
> ```
> 
> I'm just hoping that disconnection is automatic (or at least good enough) and won't cause any problems there. As far as I can tell, signals without connected slots will just do nothing and leave the `vector` empty (so the worst that happens is an empty list instead of a crash).

@BhaaLseN After removing the **GameTracker**'s worker thread and the `RunOnObject()` usage, we are off the hook. All good now.

Since this has grown a bit too large, I'd like to leave those other (pertinent) changes that you suggested for a follow-up pull request.

cristian64 commented on Source/Core/DolphinQt/Settings.h:
```diff
@@ -181,6 +183,7 @@ class Settings final : public QObject
   bool m_controller_state_needed = false;
   std::shared_ptr<NetPlay::NetPlayClient> m_client;
   std::shared_ptr<NetPlay::NetPlayServer> m_server;
+  GameListModel* m_game_list_model = nullptr;
```
In the end, QPointer has been used as the container, so it can be checked whether the object has been deleted.

cristian64 commented:
@BhaaLseN In the end, **QPointer** has been used instead. Having to define several signals and slots in several components was a bit convoluted, specially around things like the `RequestGameList()` signal/slot you suggested. `emit` is a fire and forget action, so you would not expect anything to be returned (or assigned to the output parameter).

BhaaLseN commented:
I'm pretty sure `emit` is synchronous, ie. it is _not_ fire&forget (but instead more like a multi-cast with 0..n handlers to be run in succession). But yeah, it was mostly an idea, not necessarily a "you must do it that way" thing.

Tilka commented:
With this PR I get an empty game list at startup. After about 3 seconds the complete list plops in. This didn't happen before.

cristian64 commented:
Hi @Tilka , thanks for giving the patch a go and the feedback.

In my tests, the list (30 games) is loaded immediately as the main window shows up. How large is your list? Platform?

The only change that I can think of that may have affected performance negatively is the thread's priority, which I decided to set to the very lowest, `QThread::IdlePriority`. That may have been a mistake. I'll change the priority to `QThread::InheritPriority` (default) and update the PR shortly.

Would you mind giving it another go?

Tilka commented:
There are currently 427 items in my game list. You could probably simulate this with a few hundred dol files. The thread priority change didn't improve the speed.

Tilka commented:
Is it possible that master is just showing the stale cached game list and updates it in the background as needed whereas you are only showing the cache once it has been validated? That would explain why master crashes for me if I close it right away (probably the crash you are fixing here).

cristian64 commented:
That is exactly what I'm looking at at the moment. Now I'm making sure that my local `gamelist.cache` file is erased. I'm also adding some `sleep()` in `GameFileCache::Update()` to mock a slow drive. So far I have reproduced the problem you're seeing (but I've seem other problem like duplicated entries in the list).

cristian64 commented:
I think I have spotted one(the?) problem. In short: you were right.

I messed the order of the lines `m_cache.Update(...)` and `m_cache.ForEach(...)` in `GameTracker::StartSlot()`. Basically they have to be swapped around, or else the cache is not really used, as you mentioned.

In **master**, the `m_cache.ForEach(...)` line is executed just after the cache has loaded, and before the `m_cache.Update(...)` line. But that is not the case in my PR.

Tilka commented:
Now the list shows up immediately, but when I close Dolphin right away I get this:

```
QThread: Destroyed while thread is still running

Thread 1 "dolphin-emu" received signal SIGABRT, Aborted.
0x00007fffef64dd7f in raise () from /usr/lib/libc.so.6
=> 0x00007fffef64dd7f <raise+271>:	48 8b 8c 24 08 01 00 00	mov    rcx,QWORD PTR [rsp+0x108]
(gdb) bt
#0  0x00007fffef64dd7f in raise () at /usr/lib/libc.so.6
#1  0x00007fffef638672 in abort () at /usr/lib/libc.so.6
#2  0x00007ffff46127fc in  () at /usr/lib/libQt5Core.so.5
#3  0x00007ffff4613055 in  () at /usr/lib/libQt5Core.so.5
#4  0x0000555555779d84 in GameListModel::~GameListModel() (this=0x5555579b1060, __in_chrg=<optimized out>) at /usr/include/qt/QtCore/qlist.h:75
#5  0x0000555555779d84 in GameListModel::~GameListModel() (this=0x5555579b1060, __in_chrg=<optimized out>) at ../Source/Core/DolphinQt/GameList/GameListModel.h:25
#6  0x00007ffff48317ab in QObjectPrivate::deleteChildren() () at /usr/lib/libQt5Core.so.5
#7  0x00007ffff7aa328f in QWidget::~QWidget() () at /usr/lib/libQt5Widgets.so.5
#8  0x00005555557187e9 in GameList::~GameList() (this=0x5555574b0930, __in_chrg=<optimized out>) at /usr/include/c++/8.2.1/bits/atomic_base.h:303
#9  0x0000555555718879 in GameList::~GameList() (this=0x5555574b0930, __in_chrg=<optimized out>) at ../Source/Core/DolphinQt/GameList/GameList.cpp:168
#10 0x00007ffff48317ab in QObjectPrivate::deleteChildren() () at /usr/lib/libQt5Core.so.5
#11 0x00007ffff7aa328f in QWidget::~QWidget() () at /usr/lib/libQt5Widgets.so.5
#12 0x00007ffff7aa34aa in QWidget::~QWidget() () at /usr/lib/libQt5Widgets.so.5
#13 0x00007ffff48317ab in QObjectPrivate::deleteChildren() () at /usr/lib/libQt5Core.so.5
#14 0x00007ffff7aa328f in QWidget::~QWidget() () at /usr/lib/libQt5Widgets.so.5
#15 0x00007ffff7c0f67a in QStackedWidget::~QStackedWidget() () at /usr/lib/libQt5Widgets.so.5
#16 0x00007ffff48317ab in QObjectPrivate::deleteChildren() () at /usr/lib/libQt5Core.so.5
#17 0x00007ffff7aa328f in QWidget::~QWidget() () at /usr/lib/libQt5Widgets.so.5
#18 0x000055555567ff59 in MainWindow::~MainWindow() (this=0x7fffffffe1a0, __in_chrg=<optimized out>) at /usr/include/c++/8.2.1/bits/unique_ptr.h:347
#19 0x0000555555652259 in main(int, char**) (argc=<optimized out>, argv=<optimized out>) at ../Source/Core/DolphinQt/Main.cpp:217
```

cristian64 commented:
Right... I don't think there is much to do here (but I may be wrong, as your stacktrace is not giving us all the info).

In `GameTracker::~GameTracker()` the thread is stopped. If it does not return within 5 seconds (arbitrary value), it attempts to terminate the thread (I believe that action does nothing on Linux; it does terminate the thread on Windows, but, again, I may be wrong).

The thread is potentially blocked by one of the cache functions, which you reckon may take several seconds to complete... If it takes longer than 5, bad things are going to happen... Ideally, we could try to identify what function of the cache is taking long (is it `Update()`? is it `ForEach()`? is it `Load()`?) and pass a flag by reference, so we can try cancel the slow-function's internal loop, therefore ensuring that shutting Dolphin down is very responsive.

Another option would be to increment the timeout from 5 seconds to 10 seconds... That should be a good value for everybody... :trollface:

cristian64 commented:
I have added the aforementioned flag to `GameFileCache::Update()`, since that is the one that does the bulk of the work. Also, for completeness, I've added the flag to `GameFileCache::UpdateAdditionalMetadata()`.

In theory, you should be able to close Dolphin at any time (even when the cache is still being built up for the first time), and it should react quickly, without crashing, and without leaking the **GameListModel** or the **GameTracker**.

Tilka commented:
Seems to work well now.

cristian64 commented:
Just updated the PR with a few changes:
- Rebased on **master**.
- Removed the _Slot_ suffix from certain functions names. Instead, an _On_ prefix has been added: `StartSlot() -> OnStart()`. This matches Dolphin's style a bit better.
- Changed `bool* aborted` to `const bool* aborted`, as it is an optional, read-only parameter.

cristian64 commented:
Rebased and sanity-checked that everything is still working correctly.

@Tilka Is this still under your radar?

cristian64 commented:
Rebased, addressed conflicts, and sanity-checked that functionality is still working correctly.

cristian64 commented:
Rebased, addressed conflicts, and sanity-checked that functionality is still working correctly (i.e. Dolphin no longer crashes if closed when games are being downloaded).

cristian64 commented:
Rebased, addressed conflicts, and sanity-checked that functionality is still working correctly (i.e. Dolphin no longer crashes if closed when games are being downloaded).

cristian64 commented:
@Tilka @BhaaLseN @Techjar  Hey guys, pinging you one more time to see if I can get some closure here.

I've updated the PR one more time. The PR is still relevant: Dolphin crashes on shutdown if a game file in the game list is growing.

Following the reproduction steps in the PR's description, this is the stacktrace as of 5.0-11327 (ba8b5bea11596c6ed2cb710dd35796ae750f6bd7):

```
#0  0x0000555555687583 in bool Config::Get<bool>(Config::ConfigInfo<bool> const&) ()
#1  0x00005555559de317 in UICommon::GameFile::CustomCoverChanged() ()
#2  0x00005555559e23e1 in UICommon::GameFileCache::UpdateAdditionalMetadata(std::shared_ptr<UICommon::GameFile>*) [clone .constprop.210] ()
#3  0x00005555559e33a3 in UICommon::GameFileCache::AddOrGet(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool*) ()
#4  0x00005555557599ab in GameTracker::LoadGame(QString const&) [clone .part.200] ()
#5  0x000055555575c483 in GameTracker::UpdateFileInternal(QString const&) ()
#6  0x0000555555761831 in std::_Function_handler<void (GameTracker::Command), GameTracker::Command(QObject*)::{lambda(GameTracker::Command)#3}>::_M_invoke(std::_Any_data const&, GameTracker::Command&&) ()
#7  0x0000555555761b89 in Common::WorkQueueThread<GameTracker::Command>::ThreadLoop() ()
#8  0x00007ffff09eb66f in  () at /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#9  0x00007ffff0cbe6db in start_thread (arg=0x7fffc09f4700) at pthread_create.c:463
#10 0x00007ffff02c088f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

JosJuice commented:
@dolphin-emu-bot rebuild

cristian64 commented:
Sorry, I failed to run the linter against the new Clang-Format 7 rules. I have addressed the warning in the same way this kind of issues were addressed in ff972e3.

BhaaLseN commented on Source/Core/DolphinQt/NetPlay/NetPlayDialog.cpp:
```diff
@@ -1019,10 +1018,13 @@ bool NetPlayDialog::IsRecording()
 {
   std::optional<std::string> path = RunOnObject(this, [this, &game] {
-    for (int i = 0; i < m_game_list_model->rowCount(QModelIndex()); i++)
+    if (m_game_list_model)
```
Is it even necessary to `RunOnObject` here? Or is there a chance that `m_game_list_model` would test ok before the call but fail inside?
Thinking about whether `if (!m_game_list_model) return "";` could be outside the lambda, _before_ calling `RunOnObject`

cristian64 replied:
Good question. `NetPlayDialog::FindGame()` is a public function. Technically, it could be invoked from a non-UI thread, therefore, `RunOnObject()` is required; I don't think we can remove it. Without `RunOnObject()`, it wouldn't be safe to access UI components (i.e. `m_game_list_model` in this case). At least I think that was the intention of the original writer (spycrap).

So, I think we need `if (m_game_list_model)` inside the lambda. We could also have it outside the lambda as you suggested, but it's unlikely `m_game_list_model` will be `nullptr` at that time. The user would have to start a new NetPlay game and close Dolphin quickly in order to hit the sweet gap in which the GameListModel could be destroyed before finding the game in the list. It's extremely unlikely if at all possible, and, for correctness, we would still need to check inside the lambda to be safe.

BhaaLseN commented on Source/Core/DolphinQt/NetPlay/NetPlayDialog.cpp:
```diff
@@ -1035,10 +1037,13 @@ std::shared_ptr<const UICommon::GameFile> NetPlayDialog::FindGameFile(const std:
   std::optional<std::shared_ptr<const UICommon::GameFile>> game_file =
       RunOnObject(this, [this, &game] {
-        for (int i = 0; i < m_game_list_model->rowCount(QModelIndex()); i++)
+        if (m_game_list_model)
```
Same down here; depending on what your response to the first one is.

cristian64 commented:
I guess I can rebase once more.

cristian64 commented:
Rebased.

I've sanity-checked that functionality is still working correctly (i.e. Dolphin no longer crashes if closed when games are being downloaded).

The crash is still reproducible in Dolphin 5.0-11836 (current). Stacktrace:

```
#0  0x000055555568f893 in bool Config::Get<bool>(Config::ConfigInfo<bool> const&) ()
#1  0x00005555559e8faf in UICommon::GameFile::DownloadDefaultCover() ()
#2  0x00005555559ec2ca in UICommon::GameFileCache::UpdateAdditionalMetadata(std::shared_ptr<UICommon::GameFile>*) [clone .constprop.210] ()
#3  0x00005555559ed2a3 in UICommon::GameFileCache::AddOrGet(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool*) ()
#4  0x000055555576a9db in GameTracker::LoadGame(QString const&) [clone .part.200] ()
#5  0x000055555576d493 in GameTracker::UpdateFileInternal(QString const&) ()
#6  0x00005555557727f1 in std::_Function_handler<void (GameTracker::Command), GameTracker::Command(QObject*)::{lambda(GameTracker::Command)#3}>::_M_invoke(std::_Any_data const&, GameTracker::Command&&) ()
#7  0x0000555555772b49 in Common::WorkQueueThread<GameTracker::Command>::ThreadLoop() ()
#8  0x00007ffff07e06df in  () at /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#9  0x00007ffff0ab36db in start_thread (arg=0x7fffaffff700) at pthread_create.c:463
#10 0x00007ffff00b588f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

JosJuice commented:
@dolphin-emu-bot rebuild

cristian64 commented:
Rebased one more time. Hoping to get some closure on this.

This crash is still a problem in Dolphin 5.0-12312 (Linux).

**Reproduction steps:**
- Launch Dolphin.
- Start transferring some game files to one of the directories monitored by Dolphin.
- Quit Dolphin while the game list is being periodically updated.
- Dolphin will (potentially) crash with something similar to:
```
#0  0x00005555556fe722 in std::_Rb_tree<Config::Location, std::pair<Config::Location const, std::optional<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::_Select1st<std::pair<Config::Location const, std::optional<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, std::less<Config::Location>, std::allocator<std::pair<Config::Location const, std::optional<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >::_M_begin() const (this=0x10) at /usr/include/c++/8/bits/stl_tree.h:748
#1  0x00005555556fe5cb in std::_Rb_tree<Config::Location, std::pair<Config::Location const, std::optional<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::_Select1st<std::pair<Config::Location const, std::optional<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, std::less<Config::Location>, std::allocator<std::pair<Config::Location const, std::optional<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >::find(Config::Location const&) const (this=0x10, __k=...) at /usr/include/c++/8/bits/stl_tree.h:2559
#2  0x00005555556fe40f in std::map<Config::Location, std::optional<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::less<Config::Location>, std::allocator<std::pair<Config::Location const, std::optional<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >::find(Config::Location const&) const (this=0x10, __x=...) at /usr/include/c++/8/bits/stl_map.h:1194
#3  0x00005555556fe250 in Config::Layer::Get<bool>(Config::Location const&) const (this=0x0, location=...) at /w/dolphin/Source/Core/Common/Config/Layer.h:114
#4  0x00005555556fe196 in Config::Layer::Get<bool>(Config::Info<bool> const&) const (this=0x0, config_info=...) at /w/dolphin/Source/Core/Common/Config/Layer.h:108
#5  0x00005555556fe055 in Config::Get<bool>(Config::Info<bool> const&) (info=...) at /w/dolphin/Source/Core/Common/Config/Config.h:53
#6  0x0000555555c1395e in UICommon::GameFile::CustomCoverChanged() (this=0x7fffa43d6f80) at /w/dolphin/Source/Core/UICommon/GameFile.cpp:166
#7  0x0000555555c1a054 in UICommon::GameFileCache::UpdateAdditionalMetadata(std::shared_ptr<UICommon::GameFile>*) (this=0x555557ebad28, game_file=0x7fffa41eb940) at /w/dolphin/Source/Core/UICommon/GameFileCache.cpp:176
#8  0x0000555555c19a4c in UICommon::GameFileCache::AddOrGet(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool*) (this=0x555557ebad28, path="/home/cristian/Downloads/GameCube/Resident Evil 4 (Europe) (En,Fr,De,Es,It) (Disc 1).iso", cache_changed=0x7fffafff529f)
    at /w/dolphin/Source/Core/UICommon/GameFileCache.cpp:84
#9  0x0000555555859d0a in GameTracker::LoadGame(QString const&) (this=0x555557ebac00, path=...) at /w/dolphin/Source/Core/DolphinQt/GameList/GameTracker.cpp:336
#10 0x0000555555859a12 in GameTracker::UpdateFileInternal(QString const&) (this=0x555557ebac00, file=...) at /w/dolphin/Source/Core/DolphinQt/GameList/GameTracker.cpp:295
#11 0x0000555555857d40 in GameTracker::<lambda(GameTracker::Command)>::operator()(GameTracker::Command) const (__closure=0x555557ebac20, command=...) at /w/dolphin/Source/Core/DolphinQt/GameList/GameTracker.cpp:71
#12 0x000055555585a93b in std::_Function_handler<void(GameTracker::Command), GameTracker::GameTracker(QObject*)::<lambda(GameTracker::Command)> >::_M_invoke(const std::_Any_data &, GameTracker::Command &&) (__functor=..., __args#0=...) at /usr/include/c++/8/bits/std_function.h:297
#13 0x000055555585e1eb in std::function<void (GameTracker::Command)>::operator()(GameTracker::Command) const (this=0x555557ebac20, __args#0=...) at /usr/include/c++/8/bits/std_function.h:687
#14 0x000055555585d1af in Common::WorkQueueThread<GameTracker::Command>::ThreadLoop() (this=0x555557ebac20) at /w/dolphin/Source/Core/Common/WorkQueueThread.h:69
#15 0x000055555585e3ae in std::__invoke_impl<void, void (Common::WorkQueueThread<GameTracker::Command>::*)(), Common::WorkQueueThread<GameTracker::Command>*>(std::__invoke_memfun_deref, void (Common::WorkQueueThread<GameTracker::Command>::*&&)(), Common::WorkQueueThread<GameTracker::Command>*&&) (__f=@0x55555834bc00: (void (Common::WorkQueueThread<GameTracker::Command>::*)(Common::WorkQueueThread<GameTracker::Command> * const)) 0x55555585d0ce <Common::WorkQueueThread<GameTracker::Command>::ThreadLoop()>, __t=@0x55555834bbf8: 0x555557ebac20) at /usr/include/c++/8/bits/invoke.h:73
#16 0x000055555585d26d in std::__invoke<void (Common::WorkQueueThread<GameTracker::Command>::*)(), Common::WorkQueueThread<GameTracker::Command>*>(void (Common::WorkQueueThread<GameTracker::Command>::*&&)(), Common::WorkQueueThread<GameTracker::Command>*&&) (__fn=@0x55555834bc00: (void (Common::WorkQueueThread<GameTracker::Command>::*)(Common::WorkQueueThread<GameTracker::Command> * const)) 0x55555585d0ce <Common::WorkQueueThread<GameTracker::Command>::ThreadLoop()>) at /usr/include/c++/8/bits/invoke.h:95
#17 0x000055555586078f in std::thread::_Invoker<std::tuple<void (Common::WorkQueueThread<GameTracker::Command>::*)(), Common::WorkQueueThread<GameTracker::Command>*> >::_M_invoke<0ul, 1ul>(std::_Index_tuple<0ul, 1ul>) (this=0x55555834bbf8) at /usr/include/c++/8/thread:244
#18 0x000055555586074a in std::thread::_Invoker<std::tuple<void (Common::WorkQueueThread<GameTracker::Command>::*)(), Common::WorkQueueThread<GameTracker::Command>*> >::operator()() (this=0x55555834bbf8) at /usr/include/c++/8/thread:253
#19 0x000055555586072e in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (Common::WorkQueueThread<GameTracker::Command>::*)(), Common::WorkQueueThread<GameTracker::Command>*> > >::_M_run() (this=0x55555834bbf0) at /usr/include/c++/8/thread:196
#20 0x00007ffff07c56df in  () at /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#21 0x00007ffff28c56db in start_thread (arg=0x7fffaffff700) at pthread_create.c:463
#22 0x00007ffff009aa3f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

jordan-woyak commented on Source/Core/UICommon/GameFileCache.cpp:
```diff
@@ -99,6 +99,9 @@ bool GameFileCache::Update(
   game_paths.reserve(all_game_paths.size());
   for (const std::string& path : all_game_paths)
   {
+    if (aborted && *aborted)
```
This should be an `atomic<bool>` if you want to see changes made from another thread.
The compiler is allowed to optimize away the continuous checking of a regular `bool`. 
It doesn't look like you ever pass a `nullptr`? Please pass a reference to `atomic<bool>` in that case.

cristian64 replied:
As suggested, I changed the type to `const std::atomic_bool&`.

cristian64 commented on Source/Core/DolphinQt/CheatsManager.cpp:
```diff
@@ -175,11 +175,12 @@ void CheatsManager::OnStateChanged(Core::State state)
     return;
 
-  auto* model = Settings::Instance().GetGameListModel();
+  if (m_game_list_model)
```
I've fixed it here in `CheatsManager.cpp`, and in `GameListDialog.cpp`.

I've now confirmed that the **Cheats Manager** dialog correctly displays tabs, and the codes if present. Also, I have verified that the **Select a game** dialog, used for switching games after the NetPlay session has been listed, also displays the list of games now.

Thanks for catching this; it would have been even more embarrassing.

jordan-woyak commented on Source/Core/DolphinQt/CheatsManager.cpp:
```diff
@@ -175,11 +175,12 @@ void CheatsManager::OnStateChanged(Core::State state)
     return;
 
-  auto* model = Settings::Instance().GetGameListModel();
+  if (!m_game_list_model)
```
I should have asked before.
In what situation might `m_game_list_model` be null here?
If that's actually possible it might deserve an explanation in a comment.

cristian64 replied:
Actually, this is a good question. Since they were pointers, I felt bound to check that the pointer is not null. It seems we can use const references instead.

There is no real reason to have the GameListModel dynamically allocated in the GameList constructor; it can simply be an object. Then, instead of sharing pointers to the object, const references can be used. By using const references, we are asserting that the GameListModel will outlive the components using it, which is always the case.

I will be adding a new commit for this, but it can be squashed later.

jordan-woyak commented on Source/Core/DolphinQt/GameList/GameTracker.cpp:
```diff
@@ -46,92 +48,76 @@ GameTracker::GameTracker(QObject* parent) : QFileSystemWatcher(parent)
+
+void GameTracker::AddDirectory(const QString& dir)
+{
+  QMetaObject::invokeMethod(this, "OnAddDirectory", Q_ARG(QString, dir));
```
Instead of these `invokeMethod` calls couldn't you use signals/slots?
That would be significantly cleaner.

cristian64 replied:
I have now added a number of signals, connected to the slots, that are emitted when the public functions are invoked, guaranteeing that the slots run in the GameTracker's thread.

Basically:

- `Start()` -> `emit StartRequested()` -> `OnStartRequested()`
- `AddDirectory()` -> `emit AddDirectoryRequested()` -> `OnAddDirectoryRequested()`
- `RemoveDirectory()` -> `emit RemoveDirectoryRequested()` -> `OnRemoveDirectoryRequested()`
- `RefreshAll()` -> `emit RefreshAllRequested()` -> `OnRefreshAlltRequested()`
- `PurgeCache()` -> `emit PurgeCacheRequested()` -> `OnPurgeCacheRequested()`
- `LoadCache()` -> `emit LoadCacheRequested()` -> `OnLoadCacheRequested()`

The only reason why I was using `QMetaObject::invokeMethod()` was to avoid having to declare/connect the signals, but I agree your way is _significantly_ cleaner.

jordan-woyak commented on Source/Core/DolphinQt/GameList/GameTracker.cpp:
```diff
@@ -46,92 +48,76 @@ GameTracker::GameTracker(QObject* parent) : QFileSystemWatcher(parent)
-  m_cache_loaded_event.Set();
+  m_aborted = true;
+  QThread::quit();
+  const bool joined = QThread::wait(5000);
```
This timeout seems sloppy. In testing it does time out though. Why?
The GameTracker thread was terminated within `UpdateDirectory`.
I didn't dig much deeper but maybe more functions need to be aware of the abort variable?

cristian64 replied:
I've now added a check in `UpdateDirectory()` to bail out if aborted. Also, for solving the stall while purging, I added a separate `m_pendingTask` member, that will be set when purge or refresh all are requested; this guarantees the loop in `UpdateDirectory()` is halted.

jordan-woyak commented on Source/Core/DolphinQt/GameList/GameTracker.cpp:
```diff
@@ -341,8 +312,8 @@ void GameTracker::LoadGame(const QString& path)
 }
 
-void GameTracker::PurgeCache()
+void GameTracker::OnPurgeCache()
```
If the list is currently updating the "Purge Game List Cache" action seems to just be queued and shows no response.
This isn't good UX.
Purging the cache should abort the current refresh operation and purge the cache immediately.

cristian64 replied:
I totally oversaw this one. It is indeed a problem; very obvious if a sleep is added in `GameTracker::LoadGame()`. This issue predates this PR as well.

cristian64 replied:
As mentioned in the other comment, in order to halt the loop in `UpdateDirectory()`, I've added the `m_pendingTask` flag, that is set when purging or forcing a refresh.

cristian64 commented:
Hi @jordan-woyak , Thanks for taking the time to go through the PR. Much appreciated.

I believe I have now addressed all your comments. In my tests, I've been running with `QThread::msleep(500)` in `GameTracker::LoadGame()`, and I have not witnessed any stall either while purging or while shutting down.

In regards to the timeout (of 5000 ms) in the `GameTracker::~GameTracker()`, I guess it's about providing a sensitive value. At the end of the day, when the game list is being loaded, there are a bunch of IO operations going on (such as accessing the file system, or downloading banners/covers), which I don't think we will be able to interrupt gracefully.

The bet is that loading a single game is unlikely to take longer than 5 seconds. Surely this might not always be the case... If, for instance, the user was using a network drive, and the network adapter got disconnected, it could take long for the file system to respond. I think this warrants attempting to terminate the thread if it does not respond in 5 seconds.

Thoughts?

jordan-woyak commented:
@cristian64 Maybe you could use our `PanicYesNo` so a user can decide if they want to terminate the operation or wait it out?

cristian64 commented:
> @cristian64 Maybe you could use our `PanicYesNo` so a user can decide if they want to terminate the operation or wait it out?

As suggested, I've added the check for prompting the user whether to keep waiting. Since `PanicYesNow()` can be marked as "ignored for this session", I used `AskYesNo()` instead.

cristian64 commented:
Rebased.

cristian64 commented:
Rebased.

iwubcode commented:
Not much to add but I did some testing of this and did not see any concerns adding/removing folders from the gamelist.  I also tried adding/removing files to a watched directory and that behaved ok.

cristian64 commented:
Thanks for giving it a go, @iwubcode . After 20 months, this PR has become my nemesis.

I've just realized that the failed :x: **pr:android** check is legit. The signature of a function in `Source/Core/UICommon/GameFileCache.h` has (unnecessarily) changed, and that function is used in also in `Source/Android/jni/GameList/GameFileCache.cpp`.

(The fact that the last error/warning in the log points to a missing `fmt` library led me astray! :facepalm:)

I'll update the PR to keep the signature as it was.

cristian64 commented:
Updated to avoid modifying signature of functions in `Source/Core/UICommon/GameFileCache.h`, as it was affecting the Android build.

jordan-woyak commented:
I don't like how the "Dolphin could not find any [games]" message is shown for a second when dolphin launches before the list populates.
Ideally that message should be delayed until scanning is done.
Maybe I'm nitpicking. :P

cristian64 commented:
> I don't like how the "Dolphin could not find any [games]" message is shown for a second when dolphin launches before the list populates.
> Ideally that message should be delayed until scanning is done.
> Maybe I'm nitpicking. :P

Well spotted... Let me correct that...

(I enforced the connections to be `Qt::QueuedConnection`, thinking it'd be nicer, but that means some calls that could be invoked straight away are delayed (delayed enough to let the "could not find any games" message to be displayed).)

cristian64 commented:
> > I don't like how the "Dolphin could not find any [games]" message is shown for a second when dolphin launches before the list populates.
> > Ideally that message should be delayed until scanning is done.
> > Maybe I'm nitpicking. :P
> 
> Well spotted... Let me correct that...
> 
> (I enforced the connections to be `Qt::QueuedConnection`, thinking it'd be nicer, but that means some calls that could be invoked straight away are delayed (delayed enough to let the "could not find any games" message to be displayed).)

It turned out these changes were not as straight forward as I thought they would be...

I ended up reverting some of my previous changes, and fixing the original problem (i.e. crashes and/or deadlocks on shutdown) in a different way. (In fact, now the diff is actually smaller than before.)

So, overall, with these changes:

- Closing Dolphin while a game is being copied/downloaded should always be responsive.
- Closing Dolphin while a game is being copied/downloaded should not crash.
- Closing Dolphin while refreshing/purging the game list should always be responsive.
- Closing Dolphin while refreshing/purging the game list should not crash.
- **Refresh** button and **Purge Game List Cache** action should now be synced reliable (i.e. they become disabled when a refresh is ongoing). Previously, only the **Refresh** button would sync, in some cases, but not all cases.

cristian64 commented:
Updated to address formatting.

cristian64 commented:
- Rebased.
- Addressed conflicts.

jordan-woyak commented on Source/Core/DolphinQt/MenuBar.cpp:
```diff
@@ -506,7 +506,12 @@ void MenuBar::AddViewMenu()
+  connect(&Settings::Instance(), &Settings::GameListRefreshRequested, purge_action,
+          [purge_action] { purge_action->setEnabled(false); });
+  connect(&Settings::Instance(), &Settings::GameListRefreshCompleted, purge_action,
+          [purge_action] { purge_action->setEnabled(true); });
```
The "Purge" action isn't disabled on launch. Clicking it does nothing initially until the list completely loads then it purges of course.
You'll need to cause "GameListRefreshCompleted" to be emitted on launch though. I stopped looking for the ideal means for this to happen. You're probably more familiar with this code.

jordan-woyak replied:
Bonus points for never disabling "Purge" and just making it cancel and restart the current operation.

cristian64 replied:
Let's see if it possible to abort, dispose the pending events in the working thread, and then restart, so there is no need to enable/disable the button/action.

cristian64 replied:
Unfortunately, it was not straight forward to implement your second proposal.

For now, I addressed the problem you initially mentioned; the button/action are now disabled on launch until the GameTracker is started.

While I think it should be possible to make the button/action more responsive, it involves a bit more fiddling than I first thought.

It is trivial to flush the WorkerQueueThread object before enqueuing the purge/refresh event, but there are other IO operations that can be noticeably blocking. I artificially added some delay to some of these operations to verify that flushing the event queue is not enough; things that are currently being processed also need to be halted (when possible). This means we still need to disable/enable the button/action, or else the user might feel the lack of responsiveness (but hopefully the button/action should be re-enabled much more promptly).

I guess I will prepare a new PR _soon_.

jordan-woyak commented on Source/Core/Common/WorkQueueThread.h:
```diff
@@ -27,21 +27,34 @@ class WorkQueueThread
   }
 
-private:
+  void Cancel() { m_cancelled.Set(); }
```
I understand why you had to add `Cancel` functionality to stop additional work from piling up but I think the way you exposed it loosens the invariant of this little class too much. Now there is this potential "canceled" state where the thread is still running but no more work can be done.
The object is unusable after `Cancel` until you call `Reset` and I think it would simplify things if `Cancel` cleared and shutdown internally. I realize that makes `Cancel` a blocking call, but in its only use you call the blocking `Shutdown` anyways.
But if you really really want it this way I guess I don't care that much.

cristian64 replied:
Your suggestion is reasonable. I'll change it.

cristian64 replied:
Done.

cristian64 commented:
@jordan-woyak , Although I mentioned I'd create a separate PR for the extra improvements you suggested, I ended up adding the commit to this PR. I was not too sure when this PR was going to progress, and I didn't want to create a PR with a dependency to this PR. Arguably, the last two commits of this PR could be moved to their own PR.

With this final commit, both the **Refresh** button and the **Purge Game List Cache** action will be disabled only briefly. As soon as the refresh request is processed, they will be re-enabled, allowing the user to refresh/purge the list again mid-way.

Current behavior in Dolphin 5.0-12716 (dirty):

<img width="360" alt="Dolphin 5.0-12716 (dirty)" src="https://user-images.githubusercontent.com/1853278/94859726-49087180-0435-11eb-8b85-bbbfbcf209dd.gif"/>

Demo of this PR:

<img width="360" alt="PR 7714)" src="https://user-images.githubusercontent.com/1853278/94859784-61788c00-0435-11eb-986c-8fba7e485af6.gif"/>

(For the sake of recording the videos, I added an artificial delay, so that the list would update slowly.)

jordan-woyak reviewed (approved):
Seems to behave well now and code looks good.

cristian64 commented:
> Seems to behave well now and code looks good.

Thanks for approving and your time. Much appreciated.

cristian64 commented:
@jordan-woyak Is there anything I can do to speed things up and get this through? In the past, the PRs I worked on were merged shortly after being approved, but I have a gut feeling that this one may fall through the cracks.

cristian64 commented:
Thanks for merging. Very nice.
